### PR TITLE
Move global tracer and substitute agent-core types with API types where this is trivially possible.

### PR DESCRIPTION
**Pull request by raphw**
This change is mainly generated by IDEA code substitution, with minor adjustments. By using API types, the dependency on implementation is reduced and the required API surface becomes more visible. This change is in preparation of removing the core module as a dependency of the internal plugins.

- [x] This is something else
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)

### CODE REVIEW

**Comment by raphw** [comment]
This is mostly auto-generated. I do however wonder about one difference, and that is the change here: https://github.com/elastic/apm-agent-java/compare/main...raphw:apm-agent-java:reduce-api-exposure#diff-732acc761e4310277786f5059749ac54e3f0dcfa17d8976ca3892fa7fb7d5c52L154

The use of `TraceContext` is unique here. Is it the same as using `addLink` (`addSpanLink`)?

**Review by JonasKunz — APPROVED** [review]
Approval to allow running on CI, thorough review will follow.

File: apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelper.java [hunk 1]
```diff
@@ -143,16 +143,16 @@ public Span startJmsSendSpan(Destination destination, Message message) {
...
         if (transaction != null) {
             transaction.setFrameworkName(FRAMEWORK_NAME);
         }
         return transaction;
     }
 
-    public void makeChildOf(Transaction childTransaction, Message parentMessage) {
-        TraceContext.<Message>getFromTraceContextTextHeaders().asChildOf(childTransaction.getTraceContext(), parentMessage, JmsMessagePropertyAccessor.instance());
+    public void makeChildOf(Transaction<?> childTransaction, Message parentMessage) {
+        childTransaction.addLink(JmsMessagePropertyAccessor.instance(), parentMessage);
```
[the remark this answers is not in the document]
Comment (raphw) [reply apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelper.java:155]: I can confirm that this is the only such use within the agent project. That is why I was wondering about it while not feeling sufficiently competent to implement the change. Thanks for any feedback on this!

[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelper.java:155]: Yes, this is a unique scenario that definitely complicates stuff 😐 
And for such
It is related to the tricky nature of monitoring polling APIs for messaging systems (like `receive`)- the only thing that is interesting in these operations is what happens _after they end_ - the handling of the message they return. When instrumenting `onMessage` for example, it's easy because the instrumented method encapsulates what we are interested at. In other messaging systems we found a descent way of handling poling APIs (meaning - tracing stuff outside instrumented methods), but there is no such in JMS. So what we do is a much more dangerous thing than @JonasKunz pointed out (😬 ) - we start a transaction that encapsulates the massage handling after `receive` ends and ending the transaction when the next `receive` starts, assuming that messages are read in dedicated threads and message handling is either done on the same thread, or our cross-thread context propagation will take care of that. This is why `receiveNoWait` gets a different handling- it doesn't fit a pattern of message-reading thread.
Since the capturing of "message-handling" transactions is based on this "dangerous" heuristic, we added a way to shut it off and capture transactions of only the message receive, to at least get a minimal distributed trace and help us and users analyze or improve later. This is controlled through the [`message_polling_transaction_strategy `](https://github.com/elastic/apm-agent-java/blob/7d328022a4b3fe3e7ede19bfad1cd69f9bf738a1/apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/MessagingConfiguration.java#L36)

Bottom line:
1. these "polling transactions" you see here are off by default and I am not familiar with a place where they were needed (as far as I know our message-reading-thread heuristic worked so far, including in the Spring implementation)
2. there is no real risk of child spans, as the instrumented methods normally only block and return, but not activating makes sense
3. span link is preferable in almost any way, except from the fact that in our UI this currently won't show up in service map and dependencies table. Since this is such a corner case, I'd say- let's replace it with a span link, but please rename this helper method to something else, even `addSpanLink`

[the remark this answers is not in the document]
Comment (raphw) [reply apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelper.java]: I removed the helper method as I found it to be better readable if it is placed in the code directly.

File: CHANGELOG.asciidoc [hunk 2]
```diff
@@ -20,6 +20,7 @@ endif::[]
 
 === Unreleased
 * Introduce *apm-agent-tracer* module that contains a minimal, zero-dependency `Tracer` API. - {pull}3043[#3043]
+* Move `GlobalTracer` to API module and substitute API where this is possible without further changes.
```
Comment (JonasKunz) [inline CHANGELOG.asciidoc:23]: We usually use the changelog just for changes which cause a visible effect to the end user. Your changes are just refactoring, so I think they shouldn't be part of the changelog.

Can you remove your two entries here?

**Review by JonasKunz — COMMENTED** [review]
Just a few minor comments.
In addition I think the tests are currently failing and still need some attention.

File: apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/main/java/co/elastic/apm/agent/awssdk/v1/helper/SQSHelper.java [hunk 3]
```diff
@@ -164,15 +165,18 @@ protected void setMessageContext(@Nullable Message sqsMessage, @Nullable String
...
     @Nullable
     @Override
-    public Span startSpan(Request<?> request, URI httpURI, ExecutionContext context) {
+    public Span<?> startSpan(Request<?> request, URI httpURI, ExecutionContext context) {
         if (isAlreadyActive(request)) {
-            Span activeSpan = tracer.getActiveExitSpan();
-            if (activeSpan != null && SQS_TYPE.equals(activeSpan.getSubtype())) {
-                enrichSpan(activeSpan, request, request.getEndpoint(), context);
-                activeSpan.withSync(isRequestSync(request.getOriginalRequest()));
+            AbstractSpan<?> active = tracer.getActive();
```
[the remark this answers is not in the document]
Comment (raphw) [reply apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-1-plugin/src/main/java/co/elastic/apm/agent/awssdk/v1/helper/SQSHelper.java:170]: You are right, this method was only used at two locations, so I did not want to add it to the API. I also found the code more readable by doing the check inline. I refactored it to implement the old behavior.

File: apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-common/src/main/java/co/elastic/apm/agent/awssdk/common/AbstractDynamoDBInstrumentationHelper.java [hunk 4]
```diff
@@ -49,7 +49,7 @@ public void enrichSpan(Span span, R sdkRequest, URI httpURI, C context) {
         }
 
 
-        StringBuilder name = span.getAndOverrideName(AbstractSpan.PRIO_DEFAULT);
+        StringBuilder name = span.getAndOverrideName(co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_DEFAULT);
```
[the remark this answers is not in the document]
Comment (raphw) [reply apm-agent-plugins/apm-aws-sdk/apm-aws-sdk-common/src/main/java/co/elastic/apm/agent/awssdk/common/AbstractDynamoDBInstrumentationHelper.java:52]: I was planning to do this in a step of its own, there is plenty of constants that can be moved to the API.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java [hunk 5]
```diff
@@ -88,9 +89,12 @@ public static <HttpServletRequest, HttpServletResponse, ServletContext, ServletC
                 return null;
             }
 
-            ServletServiceNameHelper.determineServiceName(adapter, adapter.getServletContext(httpServletRequest), tracer);
+            ElasticApmTracer elasticApmTracer = tracer.probe(ElasticApmTracer.class);
```
[the remark this answers is not in the document]
Comment (raphw) [reply apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java:92]: You are right, the check got duplicated, also within the method, where it is better suited.

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/ApiInstrumentation.java [hunk 6]
```diff
@@ -18,12 +18,17 @@
...
-import co.elastic.apm.agent.bci.TracerAwareInstrumentation;
+import co.elastic.apm.agent.impl.Tracer;
+import co.elastic.apm.agent.sdk.ElasticApmInstrumentation;
+import co.elastic.apm.agent.tracer.GlobalTracer;
 
 import java.util.Collection;
 import java.util.Collections;
 
-public abstract class ApiInstrumentation extends TracerAwareInstrumentation {
+public abstract class ApiInstrumentation extends ElasticApmInstrumentation {
```
[the remark this answers is not in the document]
Comment (raphw) [reply apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/ApiInstrumentation.java]: I wanted to delay migrating the plugins as the API instrumentations use plenty of "unique" APIs and added therefore a static `tracer` field which unwrapped to the old API. Turns out this was a bad idea anyways as the tracer might not be initialized such that the `require` call would fail and cause the unit tests to fail, too. I moved this back and rather require the tracer implementation where needed, inline.

**Comment by raphw** [comment]
Tests should be fixed by now. Mainly missing mock configurations to reflect `require` calls properly. And of course `JmsInstrumentationIT` needed an adjustment with regards to the span link change.

**Review by JonasKunz — APPROVED** [review]
I think we are good to merge!
Side Note for the next PR: Would be great if you could use merge instead of force-push. This allows us to use the "Review changes since last review" feature from GH when doing review iterations.

**Comment by JonasKunz** [comment]
run elasticsearch-ci/docs
